sigmastate/src/main/scala/sigmastate/serialization/ContractTemplateSerializer.scala

```diff
@@ -0,0 +1,144 @@
+    val w = SigmaSerializer.startWriter()
+    assert(contractTemplate.isRightParsed, s"Right parsed ContractTemplate expected: $contractTemplate")
+    ValueSerializer.serialize(contractTemplate.expressionTree.right.get, w)
+    w.toBytes
+  }
+
+  /** Default serialization of ContractTemplate.
+    * Doesn't apply any transformations and guarantee to preserve original
+    * structure after deserialization. */
+  def serializeContractTemplate(contractTemplate: ContractTemplate): Array[Byte] = {
+    val res = contractTemplate.expressionTree match {
+      case Left(UnparsedContractTemplate(bytes, _)) => bytes.array
+      case _ =>
+        val w = SigmaSerializer.startWriter()
+        serializeHeaders(contractTemplate, w)
+        val expressionBytes = serializeWithoutSize(contractTemplate)
+        w.putUInt(expressionBytes.length)
+        w.putBytes(expressionBytes)
+        w.toBytes
+    }
+    res
+  }
+
+  /** Default deserialization of ContractTemplate (should be inverse to `serializeContractTemplate`).
+    * Doesn't apply any transformations to the values. */
+  def deserializeContractTemplate(bytes: Array[Byte]): ContractTemplate = {
+    val r = SigmaSerializer.startReader(bytes)
+    deserializeContractTemplate(r)
+  }
+
+  private[sigmastate] def deserializeContractTemplate(r: SigmaByteReader): ContractTemplate = {
+    val startPos = r.position
+    val treeVersion = r.getOption(r.getUByte().toByte)
+
+    val nameLength = r.getUInt().toInt
+    val name = new String(r.getBytes(nameLength), StandardCharsets.UTF_8)
+
+    val descriptionLength = r.getUInt().toInt
+    val description = new String(r.getBytes(descriptionLength), StandardCharsets.UTF_8)
+
+    val nConstants = r.getUInt().toInt
+    val constTypes: IndexedSeq[SType] = {
+      if (nConstants > 0) {
+        // HOTSPOT:: allocate new array only if it is not empty
+        val res = safeNewArray[SType](nConstants)
+        cfor(0)(_ < nConstants, _ + 1) { i =>
+          res(i) = TypeSerializer.deserialize(r)
+        }
+        res
+      } else {
+        SType.EmptySeq
+      }
+    }
+    val constValues: Option[IndexedSeq[Option[SType#WrappedType]]] = r.getOption((() => {
+      if (nConstants > 0) {
+        // HOTSPOT:: allocate new array only if it is not empty
+        val res = safeNewArray[Option[SType#WrappedType]](nConstants)
+        cfor(0)(_ < nConstants, _ + 1) { i =>
+          res(i) = r.getOption((() => DataSerializer.deserialize(constTypes(i), r))())
+        }
+        res
+      } else {
+        Array.empty[Option[SType#WrappedType]]
+      }
+    })())
+
+    val nParameters = r.getUInt().toInt
+    val parameters: IndexedSeq[Parameter] = {
+      if (nParameters > 0) {
+        val res = safeNewArray[Parameter](nParameters)
+        cfor(0)(_ < nParameters, _ + 1) { i =>
+          res(i) = ParameterSerializer.deserialize(r)
+        }
+        res
+      } else {
+        Parameter.EmptySeq
+      }
+    }
```

Comment:
You need to inherit SigmaSerializer (see SigmaBoolean as an example of serializer implementation).
Put serializers in companion object of the corresponding class.
Template serializer should have just 2 methods as in SigmaBoolean, there is no need to design it after ErgoTree serializers, because they are too complex because of optimization and various use cases. 
templates instead should be simple (write all or read all or throw an SerializerException)

Reply:
Sure, thanks for the pointer. It helped. Updated the code.